lib/internal/test_runner/runner.js

```diff
@@ -112,47 +121,131 @@ function getRunArgs({ path, inspectPort }) {
   return argv;
 }
 
+class FileTest extends Test {
+  #buffer = [];
+  #handleReportItem({ kind, node, level = 0, lexeme, tokens }) {
+
+    const indent = StringPrototypeRepeat(kDefaultIndent, level + 1);
+
+    const details = (diagnostic) => {
+      return (
+        diagnostic && {
+          __proto__: null,
+          yaml:
+            `${indent}  ` +
+            ArrayPrototypeJoin(diagnostic, `\n${indent}  `) +
+            '\n',
+        }
+      );
+    };
+
+    switch (kind) {
+      case TokenKind.TAP_VERSION:
+        break; // skip
+
+      case TokenKind.TAP_PLAN:
+        this.reporter.plan(indent, node.plan.end - node.plan.start + 1);
+        break;
+
+      case TokenKind.TAP_SUBTEST_POINT:
+        this.reporter.subtest(indent, node.name);
+        break;
+
+      case TokenKind.TAP_TEST_OK:
+      case TokenKind.TAP_TEST_NOTOK:
+        const { todo, skip, pass, fail } = node.test.status;
+        const testName = node.test.description;
+        const testnumber = node.test.id;
+        let directive = {};
+
+        if (todo) {
+          this.skip(node.test.reason);
+          directive = this.getTodo(node.test.reason);
+        }
+
+        if (skip) {
+          this.todo(node.test.reason);
+          directive = this.getSkip(node.test.reason);
+        }
+
+        if (pass) {
+          this.reporter.ok(indent, testnumber, testName, null, directive);
+        } else {
+          this.reporter.fail(indent, testnumber, testName, null, directive);
+        }
+        break;
+
+      case TokenKind.TAP_YAML_START:
+        break; // skip
+
+      case TokenKind.TAP_YAML_END:
+        if (indent === kDefaultIndent) {
+          // Ignore file top level diagnostics
+          break;
+        }
+        this.reporter.details(indent, details(node.diagnostics));
+        break;
+
+      case TokenKind.UNKNOWN:
+        // TODO(manekinekko): figure out a better way to handle invalid TAP output
```

Comment:
why the todo? I think this is a good way. only issue is `invalid` and `diagnostic` are redundant

Reply:
I wasn't sure this was the right choice. Should we proceed this way? 

@cjihrig any feedback on this?